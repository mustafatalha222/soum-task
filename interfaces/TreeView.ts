export interface ITreeNode {
  name: string;
  children?: ITreeNode[];
  checked?: boolean;
  id: string;
  description?: string;
}

export interface ITreeViewProps {
  data: ITreeNode[];
  onSelectionChange?: (selectedNodes: ITreeNode[]) => void;
}
