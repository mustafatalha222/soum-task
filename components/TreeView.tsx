import { ITreeNode, ITreeViewProps } from "@/interfaces/TreeView";
import Checkbox from "expo-checkbox";
import React, { memo } from "react";
import { View, Pressable, StyleSheet, FlatList } from "react-native";
import { Colors } from "@/constants/Colors";
import { useExpandedNodes } from "@/hooks/useExpandedNodes";
import { useCheckedState } from "@/hooks/useCheckedState";
import { ThemedText } from "./ThemedText";

const TreeView = ({ data, onSelectionChange }: ITreeViewProps) => {
  const { nodes, toggleNodeCheckbox } = useCheckedState(
    data,
    onSelectionChange
  );
  const { expandedNodes, handleToggleNode } = useExpandedNodes();

  const renderItem = ({ item }: { item: ITreeNode }) => {
    const isExpanded = expandedNodes.includes(item.id);
    return (
      <View style={styles.nodeContainer}>
        <View style={styles.row}>
          <Checkbox
            testID={`checkbox-${item.id}`}
            value={item.checked}
            onValueChange={() => toggleNodeCheckbox(item)}
          />
          <Pressable onPress={() => handleToggleNode(item.id)}>
            <View style={styles.ThemedTextContainer}>
              <ThemedText type={item.children ? "defaultSemiBold" : "default"}>
                {item.name}
              </ThemedText>
              {item.description && (
                <ThemedText style={styles.nodeDescription}>
                  {item.description}
                </ThemedText>
              )}
            </View>
          </Pressable>
        </View>
        {isExpanded && item.children && (
          <FlatList
            data={item.children}
            renderItem={renderItem}
            keyExtractor={(child) => child.id}
          />
        )}
      </View>
    );
  };

  if (!data.length)
    return (
      <ThemedText type="subtitle" style={styles.noData}>
        No Data Found
      </ThemedText>
    );
  return (
    <FlatList
      data={nodes}
      renderItem={renderItem}
      keyExtractor={(node) => node.id}
    />
  );
};

const styles = StyleSheet.create({
  nodeContainer: {
    marginLeft: 20,
  },
  row: {
    flexDirection: "row",
    alignItems: "center",
    backgroundColor: Colors.light.grey,
    padding: 4,
    borderRadius: 8,
    marginBottom: 8,
    gap: 1,
  },
  ThemedTextContainer: {
    flex: 1,
    marginLeft: 10,
  },
  nodeDescription: {
    color: "#777",
    fontSize: 12,
    lineHeight: 12,
  },
  noData: {
    margin: 20,
    textAlign: "center",
  },
});

export default memo(TreeView);
