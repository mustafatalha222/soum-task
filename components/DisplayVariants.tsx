import { ITreeNode } from "@/interfaces/TreeView";
import { StyleSheet, View } from "react-native";
import { ThemedView } from "./ThemedView";
import { ThemedText } from "./ThemedText";
import React, { memo } from "react";
import { Colors } from "@/constants/Colors";

const renderSelectedChildren = (selectedItem: ITreeNode) => {
  const selectedChildren =
    selectedItem.children?.filter((child) => child.checked) || [];
  if (!selectedChildren.length) return;
  return (
    <ThemedText key={selectedItem.id} style={styles.variant}>
      {selectedItem.name} -{" "}
      {selectedChildren.map((child) => child.name).join(", ")}
    </ThemedText>
  );
};

const renderItem = ({
  item,
  isParent,
}: {
  item: ITreeNode;
  isParent?: boolean;
}) => (
  <>
    {item.checked && (item.children || isParent) ? (
      <ThemedText style={styles.variant} key={item.id}>
        All {item.name}
      </ThemedText>
    ) : (
      <>
        {renderSelectedChildren(item)}
        {item.children && (
          <View style={styles.variantWrapper}>
            {item.children.map((child) => (
              <React.Fragment key={child.id}>
                {renderItem({ item: child })}
              </React.Fragment>
            ))}
          </View>
        )}
      </>
    )}
  </>
);

const DisplayVariants = ({ selectedItems }: { selectedItems: ITreeNode[] }) => {
  return (
    <ThemedView style={styles.container}>
      <ThemedText type="subtitle" style={styles.heading}>
        Selected Variants:
      </ThemedText>
      <View style={[styles.variantWrapper]}>
        {selectedItems.map((item) => (
          <React.Fragment key={item.id}>
            {renderItem({ item, isParent: true })}
          </React.Fragment>
        ))}
      </View>
    </ThemedView>
  );
};

const styles = StyleSheet.create({
  container: { padding: 5, minHeight: 50 },
  heading: {
    marginBottom: 5,
  },
  variantWrapper: {
    flexWrap: "wrap",
    flexDirection: "row",
    gap: 3,
    overflow: "scroll",
  },
  variant: {
    backgroundColor: Colors.light.grey,
    paddingHorizontal: 10,
    fontSize: 12,
    borderRadius: 5,
    textAlign: "center",
  },
});

export default memo(DisplayVariants);
