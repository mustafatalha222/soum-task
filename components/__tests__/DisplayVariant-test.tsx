import * as React from "react";
import renderer from "react-test-renderer";
import { render, fireEvent, RenderAPI } from "@testing-library/react-native";
import DisplayVariants from "../DisplayVariants";
import { ITreeNode } from "@/interfaces/TreeView";

const selectedItems = [
  {
    id: "1",
    name: "Mobile",
    checked: true,
    children: [
      { id: "2", name: "Apple", checked: true },
      { id: "3", name: "Samsung", checked: true },
      { id: "4", name: "MI", checked: true },
    ],
  },
];

const selectedSomeItems = [
  {
    id: "1",
    name: "Mobile",
    checked: false,
    children: [
      { id: "2", name: "Apple", checked: false },
      { id: "3", name: "Samsung", checked: true },
      { id: "4", name: "MI", checked: true },
    ],
  },
];

const setupComponent = (selectedItems: ITreeNode[]) => {
  return render(<DisplayVariants selectedItems={selectedItems} />);
};

const firstChild = selectedItems[0];

describe("Display Variant Cases with selected Items", () => {
  let component: RenderAPI;

  it("display text selected variants on top", () => {
    const { getByText } = setupComponent([]);
    expect(getByText(/selected variants/i)).toBeTruthy();
  });

  it("should display parent value with All", () => {
    const { getByText } = setupComponent(selectedItems);
    expect(getByText(`All ${firstChild.name}`)).toBeTruthy();
  });

  it("should display parent with selected children by comma", () => {
    const { getByText } = setupComponent(selectedSomeItems);
    expect(
      getByText(
        `${selectedSomeItems[0].name} - ${selectedSomeItems[0].children[1].name}, ${selectedSomeItems[0].children[2].name}`
      )
    ).toBeTruthy();
  });
});

describe("Display Variant Snapshot", () => {
  it(`snapshot test for Display Variant correctly`, () => {
    const tree = renderer
      .create(<DisplayVariants selectedItems={selectedItems} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
