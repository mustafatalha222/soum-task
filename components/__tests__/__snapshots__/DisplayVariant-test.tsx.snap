// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display Variant Snapshot snapshot test for Display Variant correctly 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#fff",
      },
      {
        "minHeight": 50,
        "padding": 5,
      },
    ]
  }
>
  <Text
    style={
      [
        {
          "color": "#11181C",
        },
        undefined,
        undefined,
        undefined,
        {
          "fontSize": 20,
          "fontWeight": "bold",
        },
        undefined,
        {
          "marginBottom": 5,
        },
      ]
    }
  >
    Selected Variants:
  </Text>
  <View
    style={
      [
        {
          "flexDirection": "row",
          "flexWrap": "wrap",
          "gap": 3,
          "overflow": "scroll",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#11181C",
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          {
            "backgroundColor": "#f0f0f0",
            "borderRadius": 5,
            "fontSize": 12,
            "paddingHorizontal": 10,
            "textAlign": "center",
          },
        ]
      }
    >
      All 
      Mobile
    </Text>
  </View>
</View>
`;
